File: 3.longest-substring-without-repeating-characters.java
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int i = 0, max = 0;

        HashMap<Character, Integer> dict = new HashMap<Character, Integer>();
        int len = s.length();
        if (len == 0) {
            return 0;
        } 
        // else if (len == 1) {
        //     return 1;
        // }
int j;
        for(i=0, j=0;i<len;i++ )
         {
            char a = s.charAt(i);

            if (dict.containsKey(a)) {
                // doing this max makes it possible for us to be not clear dict each time we find a match
                // move pointer just next to the place where the element has already occured
                j = Math.max(j,dict.get(a)+1);
            } 
            dict.put(a, i);
            max = i-j+1>max?i-j+1:max;



        }
        return max;

    }
   



}
// @lc code=end

File: 3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start
from collections import defaultdict


class Solution:
    def default_fill(self):
        return 0
    def lengthOfLongestSubstring(self, s: str) -> int:
        i,j = 0,0
        max_value = 0
        holder = {}
        while j <len(s):
            # holder[s[j]]+=1
            # if (j-i+1==len(holder)):
            #     if(j-i+1>max_value):
            #         max_value = j-i+1
            # else:
            #     while(j-i+1!=len(holder)and i<j):
            #         holder[s[i]]-=1
            #         if holder[s[i]]==0:
            #             holder.pop(s[i])  
            #         i+=1
            # j+=1
            if(s[j] in holder and s[j]>i):
                i = s[j]
            elif j-i+1 > max_value:
                max_value=j-i+1
            s[j]=j
        return max_value
            
            
        
# @lc code=end

